### PR TITLE
[FLINK-20850][runtime] Removing usage of CoLocationConstraints

### PR DESCRIPTION
**Pull request by XComp**
## What is the purpose of the change

The `CoLocationConstraint` class became obsolete when introducing [FLINK-18690](https://issues.apache.org/jira/browse/FLINK-18689).

This PR depends on PR #14580 covering [FLINK-20892](https://issues.apache.org/jira/browse/FLINK-20892).

## Brief change log

* The `CoLocationConstraint` was removed from the code clearing any existing member fields and updating/removing related test cases.

## Verifying this change

CI Build

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by XComp** [comment]
The refactoring is ready to be reviewed. Please also have a look at the docs changes that were necessary due to the refactoring.
@zhuzhurk It would be helpful for you to have a look as well.

**Comment by zhuzhurk** [comment]
@XComp how about we get the base PRs merged before continuing with this PR? I can see that all the base PRs are approved already.

**Comment by XComp** [comment]
> @XComp how about we get the base PRs merged before continuing with this PR? I can see that all the base PRs are approved already.

Ok, @rmetzger merged the previous PR. I rebased `master`...

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/JobGraphTest.java [hunk 1]
```diff
@@ -426,7 +426,7 @@ public void testGetSlotSharingGroups() {
     }
 
     @Test
-    public void testGetCoLocationGroupDescriptors() {
+    public void testGetCoLocationGroup() {
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/JobGraphTest.java:429]: maybe `testGetCoLocationGroups`?

**Review by zhuzhurk — APPROVED** [review]
Thanks for opening this PR. The changes look quite good to me. There are just several minor comments.
+1 to merge it once the minor comments are addressed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java [hunk 2]
```diff
@@ -425,12 +426,12 @@ public void setStrictlyCoLocatedWith(JobVertex strictlyCoLocatedWith) {
                     "Strict co-location requires that both vertices are in the same slot sharing group.");
         }
 
-        CoLocationGroup thisGroup = this.coLocationGroup;
-        CoLocationGroup otherGroup = strictlyCoLocatedWith.coLocationGroup;
+        CoLocationGroupImpl thisGroup = this.coLocationGroup;
+        CoLocationGroupImpl otherGroup = strictlyCoLocatedWith.coLocationGroup;
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java:430]: Wouldn't it be cleaner to add the `addVertex()` method to the `CoLocationGroup` interface?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 3]
```diff
@@ -1018,10 +1018,10 @@ private void setCoLocation() {
                             "Cannot use a co-location constraint without a slot sharing group");
                 }
 
-                Tuple2<SlotSharingGroup, CoLocationGroup> constraint =
+                Tuple2<SlotSharingGroup, CoLocationGroupImpl> constraint =
                         coLocationGroups.computeIfAbsent(
                                 coLocationGroupKey,
-                                k -> new Tuple2<>(sharingGroup, new CoLocationGroup()));
+                                k -> new Tuple2<>(sharingGroup, new CoLocationGroupImpl()));
```
Comment (rmetzger) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1024]: Adding the addVertex method to the interface would not require touching this test

**Review by rmetzger — COMMENTED** [review]
Except for one comment, the change seems mostly fine.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java (hunk 2 above)
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java:430]: The idea was to have a readonly interface as the colocation groups do not need to be changed after initialization.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 3 above)
Comment (XComp) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1024]: See my answer above.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java (hunk 2 above)
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java:430]: I see, this is only the initialization.

**Comment by rmetzger** [comment]
I'll merge the change now! Thanks for working on this!

**Pull request closed without merging**